src/cursor/cur_std.c

```diff
@@ -1290,6 +1290,46 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
     API_END_RET_STAT(session, ret, cursor_bound);
 }
 
+/*
+ * __wt_cursor_bounds_save --
+ *     Save cursor bounds to restore the state.
+ */
+void
+__wt_cursor_bounds_save(WT_CURSOR *cursor, WT_CURSOR_BOUNDS_STATE *state)
+{
+    WT_ITEM_SET(state->lower_bound, cursor->lower_bound);
+    WT_ITEM_SET(state->upper_bound, cursor->upper_bound);
```

Comment:
I'm wondering about memory here, and `WT_ITEM_SET` just makes a copy of the `data` pointer in the `WT_ITEM`.  But after this is called in `__curtable_bound` using the primary cursor, the primary cursor will have a `bound` call on it.  I think if the buffer needs to grow, the old buffer pointer (what is pointed to by state->lower_bound.data) could be freed.  (`__wt_buf_set`calls `__wt_buf_grow/_worker`, which may call `realloc`).

Might be possible to reproduce by using large keys and running under a sanitizer.

I think the solution could be to make a copy of the key via `__wt_buf_set`, of course you need to free it at the end of its use.  There may be performance implication, so maybe a discussion is needed.

Reply:
It seems like we could end up doing a number of memory allocations in the bound path. I think given that setting bounds on a cursor which already has bounds is likely to be an uncommon usage we can avoid saving and restoring the state if the cursor didn't have bounds already. So a `WT_CURSOR_HAS_BOUNDS` check may be useable here to avoid slowing down the common case?

Reply:
I'll use `__wt_buf_set` instead of `WT_ITEM_SET` to save the bounds to the state, then after restoring the bounds I'll free the buffer. I'll also add in the `WT_CURSOR_BOUNDS_SET` check.

Reply:
[dacef4b](https://github.com/wiredtiger/wiredtiger/pull/8265/commits/dacef4b707d570f18dc8ecf577ea34043ef7ad94)